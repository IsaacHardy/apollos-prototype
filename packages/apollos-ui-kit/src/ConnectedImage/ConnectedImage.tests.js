import React from 'react';
import renderer from 'react-test-renderer';

import Providers from '../Providers';

import ConnectedImage from '.';

describe('the ConnectedImage component', () => {
  it('should render', () => {
    const tree = renderer.create(
      <Providers>
        <ConnectedImage
          source={{
            uri: 'https://placeholdit.co/i/150x150?bg=eeeeee&fc=577084',
            width: 150,
            height: 150,
          }}
        />
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should maintain aspect ratio', () => {
    const tree = renderer.create(
      <Providers>
        <ConnectedImage
          source={{
            uri: 'https://placeholdit.co/i/150x150?bg=eeeeee&fc=577084',
            width: 150,
            height: 150,
          }}
          maintainAspectRatio
        />
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render with a minAspectRatio', () => {
    const tree = renderer.create(
      <Providers>
        <ConnectedImage // should render as short
          source={{
            uri: 'https://picsum.photos/200/200/?random',
            width: 200,
            height: 200,
          }}
          maintainAspectRatio
          minAspectRatio={2}
        />
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render with a maxAspectRatio', () => {
    const tree = renderer.create(
      <Providers>
        <ConnectedImage // should render as tall
          source={{
            uri: 'https://picsum.photos/200/200/?random',
            width: 200,
            height: 200,
          }}
          maintainAspectRatio
          maxAspectRatio={0.5}
        />
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should render between minAspectRatio and maxAspectRatio', () => {
    const tree = renderer.create(
      <Providers>
        <ConnectedImage // should render square
          source={{
            uri: 'https://picsum.photos/200/200/?random',
            width: 200,
            height: 200,
          }}
          maintainAspectRatio
          minAspectRatio={0.5}
          maxAspectRatio={1.5}
        />
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
  it('should throw an error if minAspectRatio is used without maintainAspectRatio', () => {
    console.error = jest.fn(); // eslint-disable-line no-console
    renderer.create(
      <Providers>
        <ConnectedImage // should render square
          source={{
            uri: 'https://picsum.photos/200/200/?random',
            width: 200,
            height: 200,
          }}
          minAspectRatio={1.5}
        />
      </Providers>
    );

    expect(console.error.mock.calls).toMatchSnapshot(); // eslint-disable-line no-console
  });
  it('should throw an error if maxAspectRatio is used without maintainAspectRatio', () => {
    console.error = jest.fn(); // eslint-disable-line no-console
    renderer.create(
      <Providers>
        <ConnectedImage // should render square
          source={{
            uri: 'https://picsum.photos/200/200/?random',
            width: 200,
            height: 200,
          }}
          maxAspectRatio={1.5}
        />
      </Providers>
    );

    expect(console.error.mock.calls).toMatchSnapshot(); // eslint-disable-line no-console
  });
  it('should render a loading state', () => {
    const tree = renderer.create(
      <Providers>
        <ConnectedImage
          source={{
            uri: 'https://placeholdit.co/i/150x150?bg=eeeeee&fc=577084',
            width: 150,
            height: 150,
          }}
        />
      </Providers>
    );
    expect(tree).toMatchSnapshot();
  });
});
