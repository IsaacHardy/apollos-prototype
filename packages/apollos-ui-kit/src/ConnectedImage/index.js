import React, { PureComponent } from 'react';
import FastImage from 'react-native-fast-image';
import PropTypes from 'prop-types';

import styled from '../styled';

import SkeletonImage from './SkeletonImage';

// This mirrors the File resource we get from Heighliner:
export const ImageSourceType = PropTypes.oneOfType([
  PropTypes.shape({
    uri: PropTypes.string,
    label: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
  }),
  PropTypes.string,
]);

const aspectRatioPropValidator = (props, propName, componentName) => {
  if (props[propName] === undefined) return;

  let errorMessage = '';

  if (typeof props[propName] !== 'number') {
    errorMessage = `Invalid prop \`${propName}\` of value \`${typeof props[
      propName
    ]}\` supplied to \`${componentName}\` expected type of number`;
  }

  if (!props.maintainAspectRatio) {
    errorMessage += ` Prop maintainAspectRatio is required for use with ${propName}`;
  }

  if (typeof props[propName] !== 'number' || !props.maintainAspectRatio) {
    return new Error(errorMessage); // eslint-disable-line consistent-return
  }
};

class ConnectedImage extends PureComponent {
  static propTypes = {
    source: PropTypes.oneOfType([
      PropTypes.arrayOf(ImageSourceType),
      ImageSourceType,
    ]),
    ImageComponent: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    maintainAspectRatio: PropTypes.bool,
    isLoading: PropTypes.bool,
    onLoad: PropTypes.func,
    style: PropTypes.any, // eslint-disable-line
    minAspectRatio: aspectRatioPropValidator,
    maxAspectRatio: aspectRatioPropValidator,
  };

  static defaultProps = {
    ImageComponent: FastImage,
    maintainAspectRatio: false,
  };

  constructor(props) {
    super(props);

    /* We store `this.props.source` as state so that IF our source doesn't have `width` and `height`
     * values we can fetch them via `onLoad` and rerender with them.
     */
    this.state = { source: this.validSource };
  }

  get validSource() {
    let sources = this.props.source || {};

    if (!this.props.isLoading) {
      if (!Array.isArray(sources)) sources = [sources];

      sources = sources.map((source) => {
        let sourceAsObject = source;
        if (typeof source === 'string') sourceAsObject = { uri: source };

        return sourceAsObject;
      });

      // TODO: move this to the server!
      sources = sources.map((source) => ({
        uri: (source.uri || '').replace(/^http:\/\/|^\/\//i, 'https://'),
        ...source,
      }));
    }

    return sources;
  }

  get aspectRatio() {
    const style = {};

    // determine the aspect ratio of an image based on its width and height
    if (
      this.props.maintainAspectRatio &&
      this.state.source[0] &&
      this.state.source[0].width &&
      this.state.source[0].height
    ) {
      style.aspectRatio =
        this.state.source[0].width / this.state.source[0].height;

      // account for possible min/max aspectRatio bounds
      if (this.props.minAspectRatio || this.props.maxAspectRatio) {
        const maxAspectRatio = this.props.maxAspectRatio || style.aspectRatio;
        const minAspectRatio = this.props.minAspectRatio || 0;

        style.aspectRatio = Math.max(
          Math.min(maxAspectRatio, style.aspectRatio), // == smaller of maxAspectRatio and current aspectRatio
          minAspectRatio
        ); // == larger of calculated "max" aspect ratio and the minimum aspect ratio
      }
    }

    return style;
  }

  handleOnLoad = (event) => {
    /* `onLoad` is triggered after or `ImageComponent` (default: `FastImage`) returns our requested
     * image. We only need to execute this code if the our original source object didn't have
     * `width` and `height` values which we need to calculate an `aspectRatio` with.
     */

    event.persist(); // TODO: Look into removing this.

    if (!this.state.source[0].width || !this.state.source[0].height) {
      const loadedImageProperties = event.nativeEvent;
      // the linter is insisting we use the setState update syntax here ðŸ™„
      this.setState((state) => {
        const imageSources = [...state.source];

        if (!state.source[0].width) {
          imageSources[0].width = loadedImageProperties.width;
        }

        if (!state.source[0].height) {
          imageSources[0].height = loadedImageProperties.height;
        }
        return { source: imageSources };
      });
    }

    if (this.props.onLoad) this.props.onLoad(event);
  };

  render() {
    const {
      ImageComponent,
      style,
      forceRatio,
      isLoading,
      maintainAspectRatio,
      ...otherProps
    } = this.props;

    return (
      <SkeletonImage onReady={!isLoading} forceRatio={forceRatio} style={style}>
        <ImageComponent
          source={this.state.source}
          onLoad={this.handleOnLoad}
          style={[this.aspectRatio, style]}
          {...otherProps}
        />
      </SkeletonImage>
    );
  }
}

const ConnectedImageWithBackgroundColor = styled(
  ({ theme }) => ({
    backgroundColor: theme.colors.background.inactive,
  }),
  'ui-kit.ConnectedImage.withBackgroundColor'
)(ConnectedImage);

ConnectedImageWithBackgroundColor.propTypes = ConnectedImage.propTypes;

export default ConnectedImageWithBackgroundColor;
