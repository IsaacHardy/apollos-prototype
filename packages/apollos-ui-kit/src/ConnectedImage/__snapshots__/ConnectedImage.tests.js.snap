// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ConnectedImage component should maintain aspect ratio 1`] = `
<View
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "aspectRatio": 1,
            },
            Object {
              "backgroundColor": "#A5A5A5",
            },
          ],
        ]
      }
    >
      <FastImageView
        onFastImageLoad={[Function]}
        resizeMode="cover"
        source={
          Object {
            "height": 150,
            "uri": "https://placeholdit.co/i/150x150?bg=eeeeee&fc=577084",
            "width": 150,
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`the ConnectedImage component should render 1`] = `
<View
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {},
            Object {
              "backgroundColor": "#A5A5A5",
            },
          ],
        ]
      }
    >
      <FastImageView
        onFastImageLoad={[Function]}
        resizeMode="cover"
        source={
          Object {
            "height": 150,
            "uri": "https://placeholdit.co/i/150x150?bg=eeeeee&fc=577084",
            "width": 150,
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`the ConnectedImage component should render a loading state 1`] = `
<View
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {},
            Object {
              "backgroundColor": "#A5A5A5",
            },
          ],
        ]
      }
    >
      <FastImageView
        onFastImageLoad={[Function]}
        resizeMode="cover"
        source={
          Object {
            "height": 150,
            "uri": "https://placeholdit.co/i/150x150?bg=eeeeee&fc=577084",
            "width": 150,
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`the ConnectedImage component should render between minAspectRatio and maxAspectRatio 1`] = `
<View
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "aspectRatio": 1,
            },
            Object {
              "backgroundColor": "#A5A5A5",
            },
          ],
        ]
      }
    >
      <FastImageView
        maxAspectRatio={1.5}
        minAspectRatio={0.5}
        onFastImageLoad={[Function]}
        resizeMode="cover"
        source={
          Object {
            "height": 200,
            "uri": "https://picsum.photos/200/200/?random",
            "width": 200,
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`the ConnectedImage component should render with a maxAspectRatio 1`] = `
<View
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "aspectRatio": 0.5,
            },
            Object {
              "backgroundColor": "#A5A5A5",
            },
          ],
        ]
      }
    >
      <FastImageView
        maxAspectRatio={0.5}
        onFastImageLoad={[Function]}
        resizeMode="cover"
        source={
          Object {
            "height": 200,
            "uri": "https://picsum.photos/200/200/?random",
            "width": 200,
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`the ConnectedImage component should render with a minAspectRatio 1`] = `
<View
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "aspectRatio": 2,
            },
            Object {
              "backgroundColor": "#A5A5A5",
            },
          ],
        ]
      }
    >
      <FastImageView
        minAspectRatio={2}
        onFastImageLoad={[Function]}
        resizeMode="cover"
        source={
          Object {
            "height": 200,
            "uri": "https://picsum.photos/200/200/?random",
            "width": 200,
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`the ConnectedImage component should throw an error if maxAspectRatio is used without maintainAspectRatio 1`] = `
Array [
  Array [
    "Warning: Failed prop type:  Prop maintainAspectRatio is required for use with maxAspectRatio
    in mapProps(getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage)))))",
  ],
  Array [
    "Warning: Failed prop type:  Prop maintainAspectRatio is required for use with maxAspectRatio
    in ConnectedImage (created by mapProps(ConnectedImage))
    in mapProps(ConnectedImage) (created by withPropsOnChange(mapProps(ConnectedImage)))
    in withPropsOnChange(mapProps(ConnectedImage)) (created by mapProps(withPropsOnChange(mapProps(ConnectedImage))))
    in mapProps(withPropsOnChange(mapProps(ConnectedImage))) (created by getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage)))))
    in getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage)))) (created by mapProps(getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage))))))
    in mapProps(getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage)))))
    in View
    in View
    in View
    in View
    in AnimatedComponent
    in SafeView
    in withOrientation
    in LayoutProvider (created by nest(ThemeProvider, LayoutProvider))
    in ThemeProvider (created by nest(ThemeProvider, LayoutProvider))
    in nest(ThemeProvider, LayoutProvider)",
  ],
]
`;

exports[`the ConnectedImage component should throw an error if minAspectRatio is used without maintainAspectRatio 1`] = `
Array [
  Array [
    "Warning: Failed prop type:  Prop maintainAspectRatio is required for use with minAspectRatio
    in mapProps(getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage)))))",
  ],
  Array [
    "Warning: Failed prop type:  Prop maintainAspectRatio is required for use with minAspectRatio
    in ConnectedImage (created by mapProps(ConnectedImage))
    in mapProps(ConnectedImage) (created by withPropsOnChange(mapProps(ConnectedImage)))
    in withPropsOnChange(mapProps(ConnectedImage)) (created by mapProps(withPropsOnChange(mapProps(ConnectedImage))))
    in mapProps(withPropsOnChange(mapProps(ConnectedImage))) (created by getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage)))))
    in getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage)))) (created by mapProps(getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage))))))
    in mapProps(getContext(mapProps(withPropsOnChange(mapProps(ConnectedImage)))))
    in View
    in View
    in View
    in View
    in AnimatedComponent
    in SafeView
    in withOrientation
    in LayoutProvider (created by nest(ThemeProvider, LayoutProvider))
    in ThemeProvider (created by nest(ThemeProvider, LayoutProvider))
    in nest(ThemeProvider, LayoutProvider)",
  ],
]
`;
